lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl

```diff
@@ -66,27 +58,27 @@ julia> ArrayInterfaceCore.flatten_tuples((1, (2, (3,))))
 
 ```
 """
-@inline function flatten_tuples(t::Tuple)
-    if @generated
-        texpr = Expr(:tuple)
-        for i in 1:fieldcount(t)
-            p = fieldtype(t, i)
-            if p <: Tuple
-                for j in 1:fieldcount(p)
-                    push!(texpr.args, :(@inbounds(getfield(getfield(t, $i), $j))))
-                end
-            else
-                push!(texpr.args, :(@inbounds(getfield(t, $i))))
+function flatten_tuples(t::Tuple)
+    fields = _new_field_positions(typeof(t))
+    ntuple(Val{nfields(fields)}()) do k
+        i, j = getfield(fields, k)
+        @inbounds j === 0 ? getfield(t, i) : getfield(getfield(t, i), j)
+    end
+end
+@assume_effects :total function _new_field_positions(T::DataType)
```

Comment:
If thers's no practical usage for `length(t) > 32`. The following simplifed version should be stable as well 
```julia
@inline flatten_tuples(t::Tuple) = (_flatten1(t[1])..., flatten_tuples(Base.tail(t))...)
flatten_tuples(::Tuple{}) = ()
_flatten1(x::Tuple) = x
_flatten1(x) = (x,)
```

Reply:
Unfortunately, that still gives inference errors on v1.6.

Reply:
It's strange. I do see some inference failures with `] test ArrayInterface`.
But I can't reproduce them under 1.6.7 REPL.
Then I try to run the testset under REPL:
```julia
julia> versioninfo()
Julia Version 1.6.7
Commit 3b76b25b64 (2022-07-19 15:11 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-9600KF CPU @ 3.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)

(@temp2) pkg> activate @temp2
  Activating environment at `C:\Users\MYJ\.julia\environments\temp2\Project.toml`

(@temp2) pkg> st
      Status `C:\Users\MYJ\.julia\environments\temp2\Project.toml`
  [4fba245c] ArrayInterface v6.0.21 `C:\Users\MYJ\.julia\dev\ArrayInterface`
  [2e50d22c] ArrayInterfaceBandedMatrices v0.1.2 `C:\Users\MYJ\.julia\dev\ArrayInterface\lib\ArrayInterfaceBandedMatrices`
  [5331f1e9] ArrayInterfaceBlockBandedMatrices v0.1.2 `C:\Users\MYJ\.julia\dev\ArrayInterface\lib\ArrayInterfaceBlockBandedMatrices`
  [30b0a656] ArrayInterfaceCore v0.1.15 `C:\Users\MYJ\.julia\dev\ArrayInterface\lib\ArrayInterfaceCore`
  [015c0d05] ArrayInterfaceOffsetArrays v0.1.6 `C:\Users\MYJ\.julia\dev\ArrayInterface\lib\ArrayInterfaceOffsetArrays`
  [b0d46f97] ArrayInterfaceStaticArrays v0.1.4 `C:\Users\MYJ\.julia\dev\ArrayInterface\lib\ArrayInterfaceStaticArrays`
  [dd5226c6] ArrayInterfaceStaticArraysCore v0.1.0 `C:\Users\MYJ\.julia\dev\ArrayInterface\lib\ArrayInterfaceStaticArraysCore`
  [aedffcd0] Static v0.7.6

julia> include("C:\\Users\\MYJ\\.julia\\dev\\ArrayInterface\\test\\runtests.jl")
```
All test passed though...

Reply:
If the concern is use of `@assume_effects :total` I can keep working on it. I figured it was safe since it was very similar to how it's used generated newly sized tuples for `NamedTuple` on merging and intersections.

Reply:
Well, 1.6 perform constprop much less aggressively than later version.
`flatten_tuples` itself is stable but the result failed to be proven as a const.
We can add `@pure` to force constprop on 1.6
```julia
if VERSION < v"1.7"
@inline Base.@pure flatten_tuples(t::Tuple) = (_flatten1(t[1])..., flatten_tuples(Base.tail(t))...)
else
@inline flatten_tuples(t::Tuple) = (_flatten1(t[1])..., flatten_tuples(Base.tail(t))...)
end
```
It should be safe as our compiler is able to prove that `flatten_tuples` has total effect on 1.8.

Reply:
a const-propable `_new_field_positions` on 1.6.7
```julia
_new_field_positions(@nospecialize(x::Tuple), n = static(1 + length(x)))  = 
    (_fl1(n - length(x), x[1])..., flatten_flag(Base.tail(x), n)...)
_new_field_positions(::Tuple{}, n) = ()
_fl1(n::Int, x::Tuple) = ntuple(Base.Fix1(tuple, n), Val(length(x)))
_fl1(n::Int, x) = ((n, 0),)

julia> @code_warntype flatten_flag((1,(1,2),(1,(1,2)),1))
Variables
  #self#::Core.Const(flatten_flag)
  x::Tuple{Int64, Tuple{Int64, Int64}, Tuple{Int64, Tuple{Int64, Int64}}, Int64}

Body::NTuple{6, Tuple{Int64, Int64}}
1 ─      nothing
│   %2 = Main.length(x)::Core.Const(4)
│   %3 = (1 + %2)::Core.Const(5)
│   %4 = Main.static(%3)::Core.Const(static(5))
│   %5 = (#self#)(x, %4)::Core.Const(((1, 0), (2, 1), (2, 2), (3, 1), (3, 2), (4, 0)))
└──      return %5
```

Reply:
We don't have `StaticInt` in ArrayInterfaceCore and when I tried that code without it 1.6.7 still had inference errors.

Reply:
Oh, I didn't test it thoughly. Sorry for that.
Although we can keep the original length with a `Val{N}` or a `NTuple{N,Bool}`. But I think it would be clearer to handle it in `flatten_tuple`
```julia
function flatten_tuples(t::Tuple)
    fields = _new_field_positions(t)
    ntuple(Val{nfields(fields)}()) do k
        i, j = getfield(fields, k)
        i = length(t) - i
        @inbounds j === 0 ? getfield(t, i) : getfield(getfield(t, i), j)
    end
end

_new_field_positions(::Tuple{}) = ()
@nospecialize
_new_field_positions(x::Tuple) = (_fl1(x, x[1])..., _new_field_positions(Base.tail(x))...)
_fl1(x::Tuple, x1::Tuple) = ntuple(Base.Fix1(tuple, length(x) - 1), Val(length(x1)))
_fl1(x::Tuple, x1) = ((length(x) - 1, 0),)
@specialize
```

Reply:
Well then I think we can let `_new_field_position` return a tuple of `(length(t) - ind1, ind2)` and let `flatten_tuples` do the final transform:
```julia
function flatten_tuples(t::Tuple)
    fields = _new_field_positions(t)
    ntuple(Val{nfields(fields)}()) do k
        i, j = getfield(fields, k)
        i = length(t) - i
        @inbounds j === 0 ? getfield(t, i) : getfield(getfield(t, i), j)
    end
end

_new_field_positions(::Tuple{}) = ()
@nospecialize
_new_field_positions(x::Tuple) = (_fl1(x, x[1])..., _new_field_positions(Base.tail(x))...)
_fl1(x::Tuple, x1::Tuple) = ntuple(Base.Fix1(tuple, length(x) - 1), Val(length(x1)))
_fl1(x::Tuple, x1) = ((length(x) - 1, 0),)
@specialize
```

Reply:
That works great! thanks!